### ISSUE TITLE
Nifti IO is faulty

### ISSUE THREAD

**Issue by dzenanz** [InsightSoftwareConsortium/ITK#1237]
Code to reproduce contributed by @issakomi:
```C++
#include "itkImage.h"
#include "itkImageFileWriter.h"
#include "itkImageFileReader.h"

int main(int argc, char ** argv)
{
  typedef itk::Image<float, 3> ImageType; 
  typedef itk::ImageFileWriter<ImageType> WriterType;
  typedef itk::ImageFileReader<ImageType> ReaderType;
  {
    ImageType::Pointer image = ImageType::New();
    ImageType::IndexType index;
    ImageType::SizeType size;
    ImageType::RegionType region;
    ImageType::SpacingType spacing;
    ImageType::PointType origin;
    ImageType::DirectionType direction;
    index.Fill(0);
    size[0] = 192;
    size[1] = 256;
    size[2] = 192;
    spacing[0] = 0.976562;
    spacing[1] = 0.976562;
    spacing[2] = 1;
    origin[0] = -100.04;
    origin[1] = -158.669;
    origin[2] = -101.794;
    direction(0,0) = 0.999998;
    direction(0,1) = 0.00174533;
    direction(0,2) = 0;
    direction(1,0) = -0.00174533;
    direction(1,1) = 0.999998;
    direction(1,2) = 0;
    direction(2,0) = 0;
    direction(2,1) = 0;
    direction(2,2) = 1;
    region.SetIndex(index);
    region.SetSize(size);
    try
    {
      image->SetRegions(region);
      image->Allocate();
      image->SetOrigin(origin);
      image->SetSpacing(spacing);
      image->SetDirection(direction);
    }
    catch (itk::ExceptionObject & ex)
    {
      std::cout << ex.GetDescription() << std::endl;
      return 1;
    }
    std::cout << "*** Origin " << image->GetOrigin() << std::endl;
    std::cout << "Direction\n" << image->GetDirection() << std::endl;
    WriterType::Pointer writer = WriterType::New();
    writer->SetInput(image);
    writer->SetFileName("output.nii.gz"); // MHA is OK
    try
    {
      writer->Write();
    }
    catch (itk::ExceptionObject & ex)
    {
      std::cout << ex.GetDescription() << std::endl;
      return 1;
    }
  }

  {
    ImageType::Pointer reopened;
    ReaderType::Pointer reader = ReaderType::New();
    reader->SetFileName("output.nii.gz"); // MHA is OK
    try
    {
      reader->Update();
      reopened = reader->GetOutput();
    }
    catch (itk::ExceptionObject & ex)
    {
      std::cout << ex.GetDescription() << std::endl;
      return 1;
    }
    std::cout << "*** Re-opened\nOrigin " << reopened->GetOrigin() << std::endl;
    std::cout << "Direction\n" << reopened->GetDirection() << std::endl;
  }
  return 0;
}
```
example output:
```text
*** Origin [-100.04, -158.669, -101.794]
Direction
0.999998 0.00174533 0
-0.00174533 0.999998 0
0 0 1

*** Re-opened
Origin [-100.04, -158.669, -101.794]
Direction
0.999998 0.00182698 0
-0.00182698 0.999998 0
0 0 1
```

**Comment by hjmjohnson** [comment]
The nifti file format does not save the 9 parameters of the direction cosign.  It stores a Quaternion representation.  There is a matrix to quaternion change that occurs in writing nifti files such that the representations are not identical.  

DC -> Quaternion -> DC conversion is not a lossless process. :(

**Comment by issakomi** [comment]
For fun i've created multi-frame  [SC DICOM file](https://drive.google.com/file/d/18u47TVsV2YlRawQ3YdNzbKIEZeLXh4To/view?usp=sharing) with the same geometry

```
RAI
Origin: [-100.04, -158.669, -101.794]
Direction: 
0.999998 0.00174533 0
-0.00174533 0.999998 0
0 0 1
```

and converted to Nifti and Nrrd with dcm2niix, the results are correct.

Nrrd header:

```
space: right-anterior-superior
sizes: 192 256 192
space origin: (99.6057,-90.3538,-101.794)
space directions: (-0.976561,0.00170442,0) (0.00170442,0.976561,0) (-0,0,1)
```
Opened in ITK

```
RPI
Origin: [-99.6057, 90.3538, -101.794]
Direction: 
0.999998 -0.00174533 0
-0.00174533 -0.999998 0
0 0 1
```

It is correct, dcm2niix exports in Slicer RAS space (_space: right-anterior-superior_), somewhere conversion happened, not really important - _the result is correct_, if change orientation filter to rai code RAI is applied - the geometry is exactly the same as in original, no loss.

```
RAI
Origin: [-100.04, -158.669, -101.794]
Direction: 
0.999998 0.00174533 0
-0.00174533 0.999998 0
0 0 1
```

**Comment by issakomi** [comment]
The issue can be merged with #1042, that Dicom example file above reproduce the same problem too.

**Comment by dzenanz** [comment]
Closing in favor of #1042.
